internal/helper/helper.go

```diff
@@ -27,6 +27,13 @@ func BoolPtr(b bool) *bool {
 	return &b
 }
 
+func PtrBool(b *bool) bool {
+	if b != nil {
+		return *b
+	}
+	return false
```

Comment:
I am a bit worried about this 'hidden' default value. We have all this logic to handle default values for booleans now, I suspect that if at this moment our bool-pointer is still nil, this is a bug. So I'd either not care about checking for `== nil` at all or error out, but then we'd have to do the same for the `Vcs` field as well.

Reply:
I think it depends if we consider that state functions are supposed to always work on a valid object or not.

If yes, then we consider the pointer is always defined and we can use it directly. We can then avoid using this `PtrBool()` and so avoid having this hidden default. You are right that the hidden default is an issue, especially because some of our booleans in the image definition are not `false` by default.
However, I feel like this solution is relying on a strong coupling between the image definition file parsing/validation and the state functions execution. I fear this will cause issues later when we want to re-use some state functions in another situation (like I did for the `pack` command) and the state machine was not initialized the same way.


If no, then we should protect every use of nested objects in the state machine to avoid panicking. For example, every time we do something like `classicStateMachine.ImageDef.Artifacts.Qcow2` we should make sure `ImageDef` is not nil, then `Artifacts` is not nil. To avoid polluting or code with this kind of checks, a common solution is to define getters on these objects and do something like `classicStateMachine.GetImageDef().GetArtifacts().GetQcow2()` with getters looking something like:

```go
func (c *classicStateMachine) GetImageDef() ImageDefinition {
	if x != nil {
		return x.ImageDef
	}
	return nil
}
```
This way we could also respect the default for each field, either by also updating the method when we update/define a default value, or by generating these method automatically (and using tags in structs to determine the default value).

@sil2100 let me know what you think about this.

Reply:
I agree, I think it's good practice to always check for correctness of input for states. Especially in the case as you mentioned: we already have the pack approach that doesn't go through the parsing phases.

But this is why I mentioned one of two options that I'd prefer: either not caring (which I agree is a bad way to go forward in this case) *or* erroring out. I think in cases where we have a pointer-to-bool, the state starts and the pointer is nil, then I don't think we should consider it a valid situation. We rely on certain mechanics to set default values to such variables, and we should ensure that the same is done when calling a state out-of-order, or otherwise we might be getting different behaviors in different situations. IMO this should really just be an error in this case.
What do you think?

Reply:
To deal with this particular situation (pointer to a bool) I agree that we can check the pointer is not nil and error out of the state function because the state function is not operating on a valid state machine. 

I also agree that we can consider each state function must make sure it can operate properly with the current state machine. I will keep an eye out on this matter because I think in some state functions we suppose the state machine is valid but it may not always be.

Reply:
I have proposed a fix. I am not 100% convinced this is the best way but we can improve this later.